docs/specs/consensus/README.md

```diff
@@ -0,0 +1,1124 @@
+
+### 5.3.1 Genesis checkpoints
+
+The checkpoints for genesis block `G` are initialized like this
+
+```rust
+// Set staking epoch data
+cState(G).staking_epoch_data.seed = Epoch_seed::zero
+cState(G).staking_epoch_data.start_checkpoint = State_hash::zero
+cState(G).staking_epoch_data.lock_checkpoint = State_hash::zero
+cState(G).staking_epoch_data.epoch_length = 1
+
+// Set next epoch data
+cState(G).next_epoch.data.seed = Epoch_seed::from_b58("2vaRh7FQ5wSzmpFReF9gcRKjv48CcJvHs25aqb3SSZiPgHQBy5Dt")
+cState(G).next_epoch_data.start_checkpoint = State_hash::zero
+cState(G).next_epoch_data.lock_checkpoint =  State_hash::from_b58("3NLoKn22eMnyQ7rxh5pxB6vBA3XhSAhhrf7akdqS6HbAKD14Dh1d")
+cState(G).staking_epoch_data.epoch_length = 2
+```
+
+The functions `Epoch_seed::from_b58` and `State_hash::from_b58` are provided by the `bin_prot` implementation.
+
+### 5.3.2 Short-range fork check
+
+Given two candidate chains, we can use the previous epoch data (`staking_epoch_data`) and the current epoch data (`next_epoch_data`) to determine whether the fork point is short-range or long-range.
+
+Remember that short-range forks are those where the fork point happens after the `lock_checkpoint` of the previous epoch, otherwise it is a long-range fork.  Observe, however, that the location of the previous epoch is a relative measurement from the perspective of a block.  If the candidate blocks are in different epochs, then they will each have a different current and previous epoch-- see the figure below.
+
+```text
+  ---
+e3 |     B1
+  ---
+e2 |     B2
+  ---
+e1 |
+  ---
+```
+
+In this example, `B1`'s current epoch is `e3` and its previous epoch is `e2`, but `B2`'s current and previous epochs are `e2` and `e1` respectively.  Observe that it is not possible to have a short-range fork if the blocks are more than one epoch apart (because the fork point would be beyond one of the blocks' previous epoch's lock checkpoint).
+
+On the other hand, if the blocks are in the same epoch, then both blocks will have the same previous epoch and thus we can simply check whether the blocks have the same `lock_checkpoint` in their previous epoch data (i.e. `B1.staking_epoch_data.lock_checkpoint == B2.staking_epoch_data.lock_checkpoint`).  This gives rise to the following algorithm.
+
+Given two chains `C1` and `C2` `isShortRange` outputs `true` if the fork is short-range, otherwise the fork is long-range and it outputs `false`.
+
+```rust
+fn isShortRange(C1, C2) -> bool
+{
+    // Get consensus state from top blocks of each chain
+    let S1 = cState(C1);
+    let S2 = cState(C2);
+
+    let check = | S1, S2 | {
+      if S1.epoch_count == S2.epoch_count => {
```

Comment:
This predicate could be outside of the check altogether, for clarity.

```rust
if S1.epoch_count == S2.epoch_count {
  S1.staking_epoch_data.lock_checkpoint == S2.staking_epoch_data.lock_checkpoint
} else {
  check(S1, S2) || check(S2, S1)
}
```

Reply:
Thanks - good suggestion.  Fixed!